src/Tasks/Microsoft.Common.CurrentVersion.targets

```diff
@@ -323,7 +323,8 @@ Copyright (C) Microsoft Corporation. All rights reserved.
     <!-- Example, C:\MyProjects\MyProject\bin\Debug\MyAssembly.dll -->
     <TargetPath Condition=" '$(TargetPath)' == '' ">$(TargetDir)$(TargetFileName)</TargetPath>
 
-    <TargetRefPath Condition=" '$(TargetRefPath)' == '' and '$(ProduceReferenceAssembly)' == 'true' ">$([MSBuild]::NormalizePath($(TargetDir), 'ref', $(TargetFileName)))</TargetRefPath>
+    <TargetRefPath Condition=" '$(TargetRefPath)' == '' and '$(ProduceReferenceAssembly)' == 'true' and ('$(ProduceReferenceAssemblyInBin)' == 'true' or '$([MSBuild]::AreFeaturesEnabled(17.0))' != 'true' ) ">$([MSBuild]::NormalizePath($(TargetDir), 'ref', $(TargetFileName)))</TargetRefPath>
+    <TargetRefPath Condition=" '$(TargetRefPath)' == '' and '$(ProduceReferenceAssembly)' == 'true' ">$([MSBuild]::NormalizePath($(IntermediateOutputPath), 'ref', $(TargetFileName)))</TargetRefPath>
```

Comment:
Should this one point to `refint`?

Reply:
There are two ref assemblies:

1. The one that the compiler produces as an output. This one is written to any time the compiler runs.
2. The one that other projects reference. This one is copied from the former but _only if it has changed_, allowing the better incremental build of referencing projects.

TargetRefPath is the path of the latter; IntermediateRefAssembly the former.

Reply:
@sharwell @jaredpar

Shouldn't the compiler generate the `ref` assembly only, if it detects the change (_i.e. when doing a 2nd build or the 3rd, etc…_)?
Generating every-time incurs perf cost, doesn't it? Is it possible to input the files already generated by the compiler to determine determinism? 😏

pardon my ignorance!

Reply:
> Shouldn't the compiler generate the ref assembly only, if it detects the change (i.e. when doing a 2nd build or the 3rd, etc…)?

No because that violates the areas of concern. The compiler is not a build system, it's a compiler. It's job is to take source and produce binaries (or more often errors). Once invoked the compiler should always produce the desired outputs. MSBuild, and the compiler MSBuild task, has the responsibilities of the build system and that includes up to date checks and such. 

The way this works at a high level is that the `CoreCompile` target has two outputs: the ref and the impl assembly. The core compiler itself will always produce the ref and impl assembly when invoked. The Compiler MSBuild task, which wraps the compiler, essentially looks at the ref assembly output and determines if it is equal to the previous ref assembly output (just compare the MVID). If not then it doesn't update the ref assembly output from the task. Hence the output of the `CoreCompile` target has two outputs (ref + impl) but one is up to date because it has the same timestamp (the ref assembly) and the other is not up to date (the impl assembly). This lets MSBuild effectively short circuit later evaluations that depend only on the ref assembly.

> pardon my ignorance!

Not ignorance at all. This is a tricky area to understand if you're not deeply involved. And most likely @rainersigwald is going to correct a few details I got wrong 😄

Reply:
> And most likely @rainersigwald is going to correct a few details I got wrong 😄

I've only got one small nit . . .

> The Compiler MSBuild task, which wraps the compiler, essentially looks at the ref assembly output and determines if it is equal to the previous ref assembly output (just compare the MVID).

This is actually done by a separate task that lives in Roslyn, [`CopyRefAssembly`](https://github.com/dotnet/roslyn/blob/dec4f470707b7337e2271c918e3a6531d0c71ff5/src/Compilers/Core/MSBuildTask/CopyRefAssembly.cs), which is called here:

https://github.com/dotnet/msbuild/blob/d150e93ff1f8828bb8b0b5fda64f4582d61a7e73/src/Tasks/Microsoft.Common.CurrentVersion.targets#L4524-L4534

Reply:
I was so close!